include/boost/math/distributions/dirichlet.hpp

```diff
@@ -43,350 +44,418 @@ namespace math
     {
       *result = policies::raise_domain_error<RealType>(
           function,
-          "variance argument is %1%, but must be > 0 !", i, pol);
+          "variance argument is %1%, but must be > 0 !", variance[i], pol);
       return false;
     }
   }
   return true;
 } // bool check_variance
+
+template <class RandomAccessContainer, class Policy>
+inline bool check_mean_and_variance(const char *function,
+                                    const RandomAccessContainer &mean,
+                                    const RandomAccessContainer &variance,
+                                    typename RandomAccessContainer::value_type *result,
+                                    const Policy &pol)
+{
+  return check_mean(function, mean, result, pol) && check_variance(function, variance, result, pol);
+} // bool check_mean_and_variance
+
+template <class RandomAccessContainer>
+inline typename RandomAccessContainer::value_type mvar_beta(
+    const RandomAccessContainer &alpha,
+    const typename RandomAccessContainer::value_type &b)
+{
+  // B(a1,a2,...ak) = tgamma(a1+a2+...+ak)/(tgamma(a1)*tgamma(a2)...*tgamma(ak)
+  using RealType = typename RandomAccessContainer::value_type;
+  RealType mb;
+  RealType alpha_sum = accumulate(alpha.begin(), alpha.end(), b * alpha.size());
+  for (decltype(alpha.size()) i = 0; i < alpha.size(); ++i)
+  {
+    mb *= tgamma(alpha[i] + b);
+  }
+  mb /= tgamma(alpha_sum);
+  return mb;
+} // mvar_beta
+
+template <class RandomAccessContainer>
+inline typename RandomAccessContainer::value_type alpha0(const RandomAccessContainer &alpha)
+{
+  return accumulate(alpha.begin(), alpha.end(), 0);
+}
 } // namespace dirichlet_detail
 
-template <class VectorType = std::vector<double>, class RealType = double, class Policy = policies::policy<>>
+template <class RandomAccessContainer = std::vector<double>, class Policy = policies::policy<>>
 class dirichlet_distribution
 {
+  using RealType = typename RandomAccessContainer::value_type;
+
 public:
-  dirichlet_distribution(VectorType concentration) : concentration(concentration)
+  dirichlet_distribution(RandomAccessContainer &&alpha) : m_alpha(alpha)
   {
     RealType result;
     dirichlet_detail::check_dist(
         "boost::math::dirichlet_distribution<%1%>::dirichlet_distribution",
-        concentration,
+        alpha,
         &result, Policy());
-    sum_concentration = accumulate(concentration.begin(), concentration.end(), 0);
   } // dirichlet_distribution constructor.
 
-  // Accessor functions:
-  VectorType Concentration() const
-  {
-    return concentration;
-  }
+  // Get the concentration parameters.
+  RandomAccessContainer &alpha() const { return m_alpha; }
+  // Set the concentration parameters.
+  RandomAccessContainer &alpha() { return m_alpha; }
 
-  size_t Order() const
-  {
-    return concentration.size();
-  }
+  // Get the order of concentration parameters.
+  decltype(m_alpha.size()) &order() const { return m_alpha.size(); }
```

Comment:
This shouldn't be a reference.

Reply:
We don't test routinely float128 (because it is slow and GCC only).  When the whole package is working OK for the Big Three MSVC, GCC and Clang, we might run a one-off check.  Boost.Multipreciison cpp_bin_float_quad is entirely portable (if slower) but too slow for routine CI testing.  Lets make sure that we can walk first ;-)